dali/kernels/audio/mel_scale/mel_filter_bank_cpu_impl.h

```diff
@@ -0,0 +1,209 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_AUDIO_MEL_SCALE_MEL_FILTER_BANK_CPU_IMPL_H_
+#define DALI_KERNELS_AUDIO_MEL_SCALE_MEL_FILTER_BANK_CPU_IMPL_H_
+
+#include <cmath>
+#include <memory>
+#include <vector>
+#include "dali/kernels/audio/mel_scale/mel_filter_bank_args.h"
+#include "dali/core/force_inline.h"
+
+namespace dali {
+namespace kernels {
+namespace audio {
+
+template <typename T>
+struct HtkMelScale {
+  static DALI_FORCEINLINE T hz_to_mel(T hz) {
+    // equivalent to `2595.0 * std::log10(1 + hz / 700.0)`
+    return T(1127) * std::log(T(1) + hz / T(700));
+  }
+
+  static DALI_FORCEINLINE T mel_to_hz(T mel) {
+    // equivalent to `700.0 * (std::pow(10, mel / 2595.0) - 1.0)`
+    return T(700) * (std::exp(mel / T(1127)) - T(1));
+  }
+};
+
+template <typename T>
+struct SlaneyMelScale {
+  static constexpr T fmin = 0;
+  static constexpr T fsp = 200.0 / 3.0;
+
+  static constexpr T min_log_hz = 1000.0;
+  static constexpr T min_log_mel = (min_log_hz - fmin) / fsp;
+  static constexpr T step_log = 0.068751777;  // Equivalent to std::log(6.4) / 27.0;
+
+  static DALI_FORCEINLINE T hz_to_mel(T hz) {
+    T mel = 0;
+    if (hz >= min_log_hz) {
+      // non-linear scale
+      mel = min_log_mel + std::log(hz / min_log_hz) / step_log;
+    } else {
+      // linear scale
+      mel = (hz - fmin) / fsp;
+    }
+
+    return mel;
+  }
+
+  static DALI_FORCEINLINE T mel_to_hz(T mel) {
+    T hz = 0;
+    if (mel >= min_log_mel) {
+      // non linear scale
+      hz = min_log_hz * std::exp(step_log * (mel - min_log_mel));
+    } else {
+      // linear scale
+      hz = fmin + mel * fsp;
+    }
+    return hz;
+  }
+};
+
+template <typename T>
+class MelFilterBankImpl {
+ public:
+  virtual ~MelFilterBankImpl() = default;
+  virtual void Compute(T* out, const T* in, int64_t nwindows,
+                       int64_t out_stride = -1, int64_t in_stride = -1) = 0;
+};
+
+// In the outer loop we travel at a linearly spaced frequency grid in the mel scale
+// Each triangular filter is defined by three points in this grid (left, center, right)
+// For each iteration we process a range between two mel frequencies in the grid, calculating
+// the contribution of each FFT bin to 2 triangular filters (one is in the negative slope region
+// and the other in the positive slope region), except for the first and last iteration.
+// In total, we do a single pass on every FFT bin column
+//
+// For every FFT bin we compute the weight for each filter and travel through the row, computing
+// the contributions on every window of the spectrogram (horizontal axis)
+//
+template <typename T, typename MelScale = HtkMelScale<T>>
```

Comment:
Unless you envision totally replacing this implementation based on MelScale, then this argument doesn't need to be here at all - it's only needed in constructor.

Reply:
done